dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/DefaultEventDataQueryService.java

```diff
@@ -223,9 +226,63 @@ public EventQueryParams getFromRequest(EventDataQueryRequest request, boolean an
       eventQueryParams = builder.build();
     }
 
+    validateQueryParamsForConfidentialAndSkipAnalytics(eventQueryParams);
+
     return eventQueryParams;
   }
 
+  private void validateQueryParamsForConfidentialAndSkipAnalytics(
+      EventQueryParams eventQueryParams) {
+    if (eventQueryParams.includeConfidentialOrSkipAnalyticsItems()) {
+      return;
+    }
+    Set<TrackedEntityAttribute> confidentialAttributes =
+        Stream.concat(
+                eventQueryParams.getItems().stream(), eventQueryParams.getItemFilters().stream())
+            .map(QueryItem::getItem)
+            .filter(TrackedEntityAttribute.class::isInstance)
```

Comment:
Using `instance of` is not entirely reliable when using Hibernate and proxies. Instead there is a `DimensionItemType` property on `DimensionalItemObject` items which you can use for this purpose.

```
item.getItem().getDimensionItemType()
```